src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -427,18 +427,61 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
         {
-            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var pgType))
+            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var postgresType))
                 throw new InvalidOperationException($"Couldn't find PostgreSQL type with OID {oid}");
 
-            if (!Mappings.TryGetValue(pgType.Name, out var mapping) && !Mappings.TryGetValue(pgType.FullName, out mapping))
-                throw new InvalidOperationException($"No mapping found for PostgreSQL type {pgType.DisplayName}");
+            // Try to find the postgresType in the mappings 
+            if (TryGetTypeInfoFromMapping(postgresType, out NpgsqlDbType? npgsqlDbType))
+                return (npgsqlDbType, postgresType, postgresType.DisplayName);
 
-            if (!mapping.NpgsqlDbType.HasValue)
-                throw new InvalidOperationException($"Invalid parameter type: {oid}");
+            if (postgresType is PostgresArrayType arrayType)
+            {
+                // Try to find the Elements' postgresType in the mappings 
+                if (TryGetTypeInfoFromMapping(arrayType.Element, out NpgsqlDbType? elementType))
+                    return (elementType | NpgsqlDbType.Array, postgresType, postgresType.DisplayName);
+
+                // It might be an array of an unmapped enum or composite type
+                if (arrayType.Element is PostgresEnumType || arrayType.Element is PostgresCompositeType)
+                    return (null, postgresType, postgresType.DisplayName);
+            }
+
+            // It might be an unmapped enum type
+            if (postgresType is PostgresEnumType || postgresType is PostgresCompositeType)
+                return (null, postgresType, postgresType.DisplayName);
+
+            throw new InvalidOperationException($"No mapping found for PostgreSQL type {postgresType.DisplayName}");
+        }
+
+        bool TryGetTypeInfoFromMapping(PostgresType pgType, out NpgsqlDbType? type)
+        {
+            if (!(Mappings.TryGetValue(pgType.Name, out var mapping) || Mappings.TryGetValue(pgType.FullName, out mapping)))
+            {
+                type = null;
+                return false;
+            }
+
+            if (mapping.NpgsqlDbType.HasValue)
+            {
+                type = mapping.NpgsqlDbType.Value;
+                return true;
+            }
+
+            var factoryType = mapping.TypeHandlerFactory.GetType();
+
+            if (factoryType.IsGenericType)
+            {
+                var typeDef = factoryType.GetGenericTypeDefinition();
+                if (typeDef == typeof(MappedCompositeTypeHandlerFactory<>) ||
+                    typeDef == typeof(EnumTypeHandlerFactory<>))
+                {
+                    type = null;
+                    return true;
+                }
+            }
 
-            return mapping.NpgsqlDbType.Value;
+            throw new InvalidOperationException($"Invalid parameter type: {pgType.OID}");
```

Comment:
As mentioned above, am not clear on this... It seems it will throw for PG types that have a mapping, but no NpgsqlDbType, and that aren't enum/composite. For one thing, this will be the case of any mappings added via the new external plugin API, since NpgsqlDbType cannot be updated. In other word, it's totally fine for a mapping to not have an NpgsqlDbType - we can still send parameters by populating `DataTypeName`. Even if this weren't the case, these methods aren't necessarily for use only by parameter derivation (even if they currently are) - there's no reason to limit them by throwing an exception if no NpgsqlDbType is found.

Reply:
Agreed and fixed.